### PR TITLE
[feat] `inlineCss` option

### LINKED ISSUE #962: Can CSS be inlined in the server-rendered pages ?
Currently, HTML is rendered in the response by the server, but the CSS is linked through <link> attributes, thus delaying proper page rendering until the stylesheet request is done.

Is it possible to inline CSS (or at least the 'start' stylesheet) like Next.JS does? Or is it planned?

### PR DESCRIPTION
**Pull request by untp**
fixes #962

### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by Rich-Harris** [comment]
Thanks — have updated this PR so that it's once again in a mergeable state.

In principle I'm in favour of this change (though I wonder if it should be a threshold rather than a toggle — if one page had a very large amount of CSS, you might want that page to use `<link>` rather than `<style>`.

Unfortunately there's one major problem. If you visit a page that has styles from components A, B and C, you'll have this...

```html
<style>
  /* A */
  /* B */
  /* C */
</style>
```

...then if you navigate to a page that has components B, C and D you'll get this:

```html
<style>
  /* A */
  /* B */
  /* C */
</style>

<link rel="stylesheet" href="/assets/B.css">
<link rel="stylesheet" href="/assets/C.css">
<link rel="stylesheet" href="/assets/D.css">
```

If you then navigate back to the first page, you'll load A again.

In contrast, when styles are always included as links, the first navigation will only load D, and the second navigation won't load anything at all. Aside from resulting in faster navigation and less wasted bandwidth, this also prevents a situation where the CSSOM is cluttered with duplicated selectors.

The CSS loading is done by Vite here — [here's where it checks to see if styles are already in the page](https://github.com/vitejs/vite/blob/71868579058512b51991718655e089a78b99d39c/packages/vite/src/node/plugins/importAnalysisBuild.ts#L60-L61).

I'm not sure what the best solution to this is. Perhaps we include the links as well as the `<style>`, to fool Vite, but add the [`disabled`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link#attr-disabled) attribute so that they remain inert?

**Comment by Rich-Harris** [comment]
Implemented `<link disabled>`, and took the opportunity to tidy up the page rendering code a bit since it was starting to get somewhat hairy

**Comment by untp** [comment]
Thanks for the commits.

I think using disabled attribute is the best option.

It is better to using threshold. Because there are some situations that it is needed. For example, when using cloudflare workers adapter, it is better to keeping server script size small, because cloudflare workers script can be up to 1 MB after compression. Also developer may want to reduce bandwidth with only inlining small css files.

The test is failing because of [svelte-announcer](https://github.com/sveltejs/kit/blob/master/packages/kit/src/core/create_app/index.js) style. We can inline styles of svelte-announcer as discussed in #1386. Whether inlineCss enabled or not, svelte-announcer's style should be inlined.

File: .changeset/tiny-files-smile.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+[feat] inlineCss option
```
Comment (Rich-Harris) [inline .changeset/tiny-files-smile.md:5]:
```suggestion
add inlineStyleThreshold option, below which stylesheets are inlined into the page
```

**Comment by gustavopch** [comment]
What's the unit in `inlineStyleThreshold`? Bytes? BTW, it may be too late, but it could have been named `minBytesToInlineStyle` to make that self-evident.

**Pull request merged**